pkg/csi/service/wcp/controller.go

```diff
@@ -1040,12 +1046,115 @@ func (c *controller) ValidateVolumeCapabilities(ctx context.Context, req *csi.Va
 	}, nil
 }
 
+// ListVolumes returns the mapping of the volumes and corresponding published nodes.
 func (c *controller) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 	*csi.ListVolumesResponse, error) {
+	start := time.Now()
+	volumeType := prometheus.PrometheusBlockVolumeType
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
-	log.Infof("ListVolumes: called with args %+v", *req)
-	return nil, status.Error(codes.Unimplemented, "")
+	var err error
+	if !commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ListVolumes) {
+		return nil, status.Error(codes.Unimplemented, "")
+	}
+	controllerListVolumeInternal := func() (*csi.ListVolumesResponse, string, error) {
+		log.Debugf("ListVolumes called with args %+v, expectedStartingIndex %v", *req, expectedStartingIndex)
+		k8sVolumeIDs := commonco.ContainerOrchestratorUtility.GetAllVolumes()
+
+		startingIdx := 0
+		if req.StartingToken != "" {
+			startingIdx, err = strconv.Atoi(req.StartingToken)
+			if err != nil {
+				log.Errorf("Unable to convert startingToken from string to int err=%v", err)
+				return nil, csifault.CSIInvalidArgumentFault, status.Error(codes.InvalidArgument,
+					"startingToken not a valid integer")
+			}
+		}
+
+		// If the startingIdx is zero or not equal to the expectedStartingIndex, then
+		// it means the listVolume request is a new one and not part of a previous
+		// request, so fetch the volumes from CNS again
+		if startingIdx == 0 || startingIdx != expectedStartingIndex {
```

Comment:
Why would you need to fetch the query volumes again and again if `startingIdx != expectedStartingIndex` ?
Didn't quite get it.